File: ios/OfflineDownloader/OfflineVideoPlugin.swift
import Foundation
import AVFoundation
import react_native_video

@objc(OfflineVideoPlugin)
class OfflineVideoPlugin: RNVAVPlayerPlugin {
    
    private var _playerRateChangeObserver: NSKeyValueObservation?
    private var _playerCurrentItemChangeObserver: NSKeyValueObservation?
    private var _playerItemStatusObserver: NSKeyValueObservation?
    
    // MARK: - Playback Mode
    @objc enum PlaybackMode: Int {
        case online = 0
        case offline = 1
        
        var description: String {
            switch self {
            case .online: return "ONLINE"
            case .offline: return "OFFLINE"
            }
        }
    }

    // MARK: - Singleton Pattern
    private static var instance: OfflineVideoPlugin?
    private static var playbackMode: PlaybackMode = .online
    
    private var contentDownloadCache: [String: Bool] = [:]
    private let cacheTimeout: TimeInterval = 2.0
    private var lastCacheCleanup: TimeInterval = 0
    private var cacheHitCount: [String: Int] = [:]
    private let maxCacheSize = 1000
    private let cacheCleanupInterval: TimeInterval = 600.0
    
    // MARK: - Singleton Methods
    @objc static func setPlaybackMode(_ mode: Int) {
        let newMode = PlaybackMode(rawValue: mode) ?? .online
        let oldMode = playbackMode
        playbackMode = newMode
        
        getInstance().contentDownloadCache.removeAll()
        print("🎯 Playback mode changed: \(oldMode.description) → \(newMode.description)")
    }
    
    @objc static func getPlaybackMode() -> Int {
        return playbackMode.rawValue
    }
    
    @objc static func getInstance() -> OfflineVideoPlugin {
        if instance == nil {
            instance = OfflineVideoPlugin()
        }
        return instance!
    }

    // MARK: - Plugin Registration
    override init() {
        super.init()
        OfflineVideoPlugin.instance = self
        ReactNativeVideoManager.shared.registerPlugin(plugin: self)
        print("✅ OfflineVideoPlugin registered successfully with ReactNativeVideoManager")
    }
       
    deinit {
        cleanupObservers()
        ReactNativeVideoManager.shared.unregisterPlugin(plugin: self)
        print("🗑️ OfflineVideoPlugin deinitialized")
    }

    // MARK: - JS-exposed methods
    @objc func isContentDownloaded(_ uri: String) -> Bool {
        return isContentDownloadedInternal(uri)
    }

    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }

    // MARK: - AVPlayer hook
    override func overridePlayerAsset(
        source: VideoSource,
        asset: AVAsset
    ) async -> OverridePlayerAssetResult? {

        guard let urlAsset = asset as? AVURLAsset else { return nil }
        let uri = urlAsset.url.absoluteString

        print("🔍 Checking DataSource override for: \(uri)")
        print("🎯 Playback mode: \(OfflineVideoPlugin.playbackMode.description)")

        switch OfflineVideoPlugin.playbackMode {
        case .online:
            print("🌐 ONLINE mode - using default DataSource for: \(uri)")
            return nil
            
        case .offline:
            let startTime = Date().timeIntervalSince1970
            let isCached = isContentCached(uri)
            let checkDuration = Date().timeIntervalSince1970 - startTime
                    
            print("⚡ Cache check took \(Int(checkDuration * 1000))ms")
            if isCached {
                print("📱 OFFLINE mode - using cached DataSource for: \(uri)")
                
                if let localURL = OfflineVideoRegistry.shared?.getLocalUrl(for: uri) {
                    let offlineAsset = AVURLAsset(url: localURL)
                                
                    return OverridePlayerAssetResult(type: .full, asset: offlineAsset)
                }
            } else {
                print("⚠️ OFFLINE mode - content not cached, using default: \(uri)")
            }
            return nil
        }
    }

    // MARK: - Cache Management
    
    func removeDownloadFromCache(_ downloadId: String) {
        print("🧹 Removing \(downloadId) from plugin cache")
        
        let keysToRemove = contentDownloadCache.keys.filter { uri in
            isUriRelatedToDownload(uri: uri, downloadId: downloadId)
        }
        
        keysToRemove.forEach { key in
            contentDownloadCache.removeValue(forKey: key)
            print("🗑️ Removed plugin cache key: \(key)")
        }
        
        print("✅ Removed \(keysToRemove.count) plugin cache entries for: \(downloadId)")
    }
    
    func isContentDownloadedInternal(_ uri: String) -> Bool {
        if let cached = contentDownloadCache[uri] {
            cacheHitCount[uri] = (cacheHitCount[uri] ?? 0) + 1
            return cached
        }
        
        let currentTime = Date().timeIntervalSince1970
        if currentTime - lastCacheCleanup > cacheCleanupInterval || contentDownloadCache.count > maxCacheSize {
            cleanupCacheIntelligently()
            lastCacheCleanup = currentTime
        }
        
        let startTime = Date().timeIntervalSince1970
        let isDownloaded = OfflineVideoRegistry.shared?.isContentCached(url: uri) ?? false
        let duration = Date().timeIntervalSince1970 - startTime
            
        print("⚡ Fast registry check took \(Int(duration * 1000))ms for: \(uri)")
            
        contentDownloadCache[uri] = isDownloaded
        cacheHitCount[uri] = 1
            
        return isDownloaded
    }
    
    private func cleanupCacheIntelligently() {
        let sortedByHits = cacheHitCount.sorted { $0.value > $1.value }
        let keepCount = min(500, sortedByHits.count)  // Keep top 500
        
        let keysToKeep = Set(sortedByHits.prefix(keepCount).map { $0.key })
        
        contentDownloadCache = contentDownloadCache.filter { keysToKeep.contains($0.key) }
        cacheHitCount = cacheHitCount.filter { keysToKeep.contains($0.key) }
        
        print("🧹 Intelligently cleaned cache: kept \(contentDownloadCache.count) items")
    }
    
    private func isContentCached(_ uri: String) -> Bool {
        if let cached = contentDownloadCache[uri] {
            return cached
        }
        
        guard let registry = OfflineVideoRegistry.shared else {
            contentDownloadCache[uri] = false
            return false
        }
        
        let startTime = Date().timeIntervalSince1970
        
        let isCached = registry.isContentCached(url: uri)
        
        let duration = Date().timeIntervalSince1970 - startTime
        print("💾 Fast cache check took \(Int(duration * 1000))ms for: \(uri)")
        
        contentDownloadCache[uri] = isCached
        return isCached
    }

    
    private func isUriRelatedToDownload(uri: String, downloadId: String) -> Bool {
        if uri == downloadId || uri.contains(downloadId) {
            return true
        }
        
        let uriContentId = extractContentId(from: uri)
        let downloadContentId = extractContentId(from: downloadId)
        
        return !uriContentId.isEmpty && !downloadContentId.isEmpty && uriContentId == downloadContentId
    }
    
    private func extractContentId(from identifier: String) -> String {
        let regex = try? NSRegularExpression(pattern: "([a-f0-9]{24})", options: [])
        let range = NSRange(location: 0, length: identifier.utf16.count)
        
        if let match = regex?.firstMatch(in: identifier, options: [], range: range) {
            let contentId = String(identifier[Range(match.range, in: identifier)!])
            if !contentId.isEmpty {
                print("🔍 Extracted content ID '\(contentId)' from: \(identifier)")
            }
            return contentId
        }
        
        return ""
    }

    // MARK: - Player lifecycle callbacks
    override func onInstanceCreated(id: String, player: AVPlayer) {
        print("📱 AVPlayer instance created: \(id)")
        setupObservers(for: player)
    }

    override func onInstanceRemoved(id: String, player: AVPlayer) {
        print("📱 AVPlayer instance removed: \(id)")
        cleanupObservers()
    }
    
    // MARK: - Observer Management
    private func setupObservers(for player: AVPlayer) {
        cleanupObservers()
        
        _playerRateChangeObserver = player.observe(
            \.rate, options: [.old], changeHandler: handlePlaybackRateChange
        )
        
        _playerCurrentItemChangeObserver = player.observe(
            \.currentItem, options: [.old], changeHandler: handleCurrentItemChange
        )
    }
    
    private func cleanupObservers() {
        _playerRateChangeObserver?.invalidate()
        _playerRateChangeObserver = nil
        
        _playerCurrentItemChangeObserver?.invalidate()
        _playerCurrentItemChangeObserver = nil
        
        _playerItemStatusObserver?.invalidate()
        _playerItemStatusObserver = nil
    }
    
    private func handlePlaybackRateChange(player: AVPlayer, change: NSKeyValueObservedChange<Float>) {
        if OfflineVideoPlugin.playbackMode == .offline && player.rate > 0 {
            print("📱 Offline content started playing")
        }
    }

    private func handlePlayerItemStatusChange(playerItem: AVPlayerItem, change: NSKeyValueObservedChange<AVPlayerItem.Status>) {
        if OfflineVideoPlugin.playbackMode == .offline {
            switch playerItem.status {
            case .readyToPlay:
                print("📱 Offline content ready to play")
            case .failed:
                if let error = playerItem.error {
                    print("❌ Offline playback failed: \(error.localizedDescription)")
                }
            case .unknown:
                print("⏳ Offline content loading...")
            @unknown default:
                print("🔄 Unknown player item status")
            }
        }
    }
    
    private func handleCurrentItemChange(player: AVPlayer, change: NSKeyValueObservedChange<AVPlayerItem?>) {
        _playerItemStatusObserver?.invalidate()
        _playerItemStatusObserver = nil
        
        guard let playerItem = player.currentItem else { return }
        
        _playerItemStatusObserver = playerItem.observe(
            \.status, options: [.new, .old], changeHandler: handlePlayerItemStatusChange
        )
        
        if OfflineVideoPlugin.playbackMode == .offline {
            if let urlAsset = playerItem.asset as? AVURLAsset {
                let isLocalFile = urlAsset.url.isFileURL
                print("📱 New offline content loaded: \(isLocalFile ? "local file" : "network")")
            }
        }
    }
}

File: android/src/main/java/com/offlinevideodownloader/OfflineVideoPlugin.kt
package com.offlinevideodownloader

import android.os.Handler
import android.os.Looper
import androidx.media3.common.MediaItem
import androidx.media3.common.StreamKey
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.DataSource
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.source.MediaSource
import com.brentvatne.exoplayer.RNVExoplayerPlugin
import com.brentvatne.react.ReactNativeVideoManager
import androidx.media3.common.util.Log
import com.brentvatne.common.api.Source
import androidx.media3.exoplayer.offline.Download
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

enum class PlaybackMode {
    ONLINE,   // Default - never use cache, always stream online
    OFFLINE   // Only use cache if available, no network requests
}

@UnstableApi
class OfflineVideoPlugin : RNVExoplayerPlugin {

    companion object {
        private const val TAG = "OfflineVideoPlugin"
        private var instance: OfflineVideoPlugin? = null

        private var playbackMode = PlaybackMode.ONLINE

        private val cacheCheckExecutor = Executors.newFixedThreadPool(3)
        private val handler = Handler(Looper.getMainLooper())

        fun cleanup() {
            try {
                cacheCheckExecutor.shutdown()
                if (!cacheCheckExecutor.awaitTermination(2, TimeUnit.SECONDS)) {
                    cacheCheckExecutor.shutdownNow()
                }
                Log.d(TAG, "✅ Cache executor cleaned up")
            } catch (e: Exception) {
                Log.w(TAG, "Warning: Error shutting down executor: ${e.message}")
            }
        }

        fun setPlaybackMode(mode: PlaybackMode) {
            val pluginInstance = getInstance()

            synchronized(this) {
                val oldMode = playbackMode
                playbackMode = mode

                try {
                    pluginInstance.contentDownloadCache.clear()  // ✅ Use pre-obtained instance
                    Log.d(TAG, "🎯 Playback mode changed: $oldMode → $mode")
                } catch (e: Exception) {
                    Log.w(TAG, "Warning: Error clearing cache during mode change: ${e.message}")
                }
            }
        }

        fun getPlaybackMode(): PlaybackMode = playbackMode

        fun getInstance(): OfflineVideoPlugin {
            if (instance == null) {
                instance = OfflineVideoPlugin()
            }
            return instance!!
        }
    }

    // ✅ ENHANCED: Better cache management
    private val contentDownloadCache = ConcurrentHashMap<String, Boolean>()
    private val cacheTimeout = 2000L
    private var lastCacheCleanup = 0L
    private val cacheCleanupInterval = 300000L // 5 minutes

    init {
        try {
            ReactNativeVideoManager.getInstance().registerPlugin(this)
            Log.d(TAG, "✅ OfflineVideoPlugin registered successfully with ReactNativeVideoManager")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to register plugin: ${e.message}")
        }
    }

    override fun onInstanceCreated(id: String, player: ExoPlayer) {
        OfflineVideoRegistry.registerPlayer(player)
        Log.d(TAG, "📱 ExoPlayer instance created: $id")
    }

    override fun onInstanceRemoved(id: String, player: ExoPlayer) {
        OfflineVideoRegistry.unregisterPlayer(player)
        Log.d(TAG, "📱 ExoPlayer instance removed: $id")
    }

    // ✅ NEW: Remove only specific download from plugin cache
    fun removeDownloadFromCache(downloadId: String) {
        synchronized(this) {
            try {
                Log.d(TAG, "🧹 Removing $downloadId from plugin cache")

                // ✅ FIXED: Only remove entries related to this download
                val keysToRemove = mutableListOf<String>()

                contentDownloadCache.keys.forEach { uri ->
                    // Check if URI is related to this download
                    if (isUriRelatedToDownload(uri, downloadId)) {
                        keysToRemove.add(uri)
                    }
                }

                keysToRemove.forEach { key ->
                    contentDownloadCache.remove(key)
                    Log.d(TAG, "🗑️ Removed plugin cache key: $key")
                }

                Log.d(TAG, "✅ Removed ${keysToRemove.size} plugin cache entries for: $downloadId")

            } catch (e: Exception) {
                Log.w(TAG, "Error removing download from plugin cache: ${e.message}")
            }
        }
    }

    // ✅ NEW: Check if URI is related to a specific download
    private fun isUriRelatedToDownload(uri: String, downloadId: String): Boolean {
        return try {
            // Direct match
            if (uri == downloadId || uri.contains(downloadId)) {
                return true
            }

            // Extract content ID from both and compare
            val uriContentId = extractContentId(uri)
            val downloadContentId = extractContentId(downloadId)

            if (uriContentId.isNotEmpty() && downloadContentId.isNotEmpty() && uriContentId == downloadContentId) {
                return true
            }

            false
        } catch (e: Exception) {
            Log.w(TAG, "Error checking URI relation: ${e.message}")
            false
        }
    }

    // ✅ NEW: Extract content ID from URI or download ID
    private fun extractContentId(identifier: String): String {
        return try {
            // For your URLs like: etvwin-s3.akamaized.net/6782084dc7036a0cfa096af2/HD_playlist.m3u8
            val regex = Regex("([a-f0-9]{24})")
            val match = regex.find(identifier)
            val contentId = match?.value ?: ""
            if (contentId.isNotEmpty()) {
                Log.d(TAG, "🔍 Extracted content ID '$contentId' from: $identifier")
            }
            contentId
        } catch (e: Exception) {
            Log.w(TAG, "Could not extract content ID from: $identifier")
            ""
        }
    }

    override fun overrideMediaDataSourceFactory(
        source: Source,
        mediaDataSourceFactory: DataSource.Factory
    ): DataSource.Factory? {
        return try {
            val uri = source.uri?.toString()
            if (uri == null) {
                Log.d(TAG, "URI is null, no override needed")
                return null
            }

            Log.d(TAG, "🔍 Checking DataSource override for: $uri")
            Log.d(TAG, "🎯 Playback mode: $playbackMode")

            // ✅ FIXED: Respect playback mode
            when (playbackMode) {
                PlaybackMode.ONLINE -> {
                    // ✅ ONLINE: Never override, always use default (fast)
                    Log.d(TAG, "🌐 ONLINE mode - using default DataSource for: $uri")
                    return null
                }

                PlaybackMode.OFFLINE -> {
                    // ✅ OFFLINE: Only override if content is cached
                    val isCached = isContentCached(uri)
                    if (isCached) {
                        Log.d(TAG, "📱 OFFLINE mode - using cached DataSource for: $uri")
                        val headers = extractHeadersFromSource(source)
                        return OfflineVideoRegistry.createCacheAwareDataSourceFactory(headers)
                    } else {
                        Log.w(TAG, "⚠️ OFFLINE mode - content not cached, using default: $uri")
                        return null
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error in overrideMediaDataSourceFactory: ${e.message}")
            null
        }
    }

    // ✅ Helper method
    private fun isContentCached(uri: String): Boolean {
        return try {
            // Check in-memory cache first
            contentDownloadCache[uri]?.let { return it }

            // ✅ Quick check with timeout
            val startTime = System.currentTimeMillis()
            val isCached = checkCacheWithTimeout(uri)
            val duration = System.currentTimeMillis() - startTime

            Log.d(TAG, "💾 Cache check took ${duration}ms for: $uri")

            // Cache result to avoid future checks
            contentDownloadCache[uri] = isCached
            isCached

        } catch (e: Exception) {
            Log.e(TAG, "❌ Quick cache check failed: ${e.message}")
            false
        }
    }

    private fun checkCacheWithTimeout(uri: String): Boolean {
        return try {
            val context = OfflineVideoRegistry.getAppContext()
            if (context == null) {
                Log.w(TAG, "⚠️ App context is null, cannot check cache")
                return false
            }

            if (!OfflineVideoRegistry.isInitialized()) {
                Log.w(TAG, "⚠️ Registry not initialized")
                return false
            }

            // ✅ Use timeout to prevent ANR
            val future = cacheCheckExecutor.submit<Boolean> {
                try {
                    OfflineDataSourceProvider.getInstance(context).isContentCached(uri)
                } catch (e: Exception) {
                    Log.w(TAG, "Cache check failed in executor: ${e.message}")
                    false
                }
            }

            future.get(1000L, TimeUnit.MILLISECONDS)

        } catch (e: TimeoutException) {
            Log.w(TAG, "⏰ Cache check timed out for: $uri")
            false
        } catch (e: Exception) {
            Log.w(TAG, "❌ Cache check failed: ${e.message}")
            false
        }
    }

    override fun overrideMediaItemBuilder(
        source: Source,
        mediaItemBuilder: MediaItem.Builder
    ): MediaItem.Builder? {
        return try {
            val uri = source.uri?.toString()
            if (uri == null) return null

            Log.d(TAG, "🔍 Checking MediaItem override for: $uri")
            Log.d(TAG, "🎯 MediaItem playback mode: $playbackMode")

            // ✅ FIXED: Use same playback mode logic as DataSource override
            when (playbackMode) {
                PlaybackMode.ONLINE -> {
                    // ✅ ONLINE: Never override MediaItem (fast)
                    Log.d(TAG, "🌐 ONLINE mode - using default MediaItem for: $uri")
                    return null
                }

                PlaybackMode.OFFLINE -> {
                    // ✅ OFFLINE: Only override if content is cached
                    val isCached = isContentCached(uri)
                    if (isCached) {
                        Log.d(TAG, "📱 OFFLINE mode - using cached MediaItem for: $uri")

                        val streamKeys = getStreamKeys(uri)
                        if (streamKeys.isNotEmpty()) {
                            Log.d(TAG, "🔑 Setting ${streamKeys.size} stream keys")
                            mediaItemBuilder.setStreamKeys(streamKeys)
                        }
                        return mediaItemBuilder
                    } else {
                        Log.w(TAG, "⚠️ OFFLINE mode - content not cached, using default: $uri")
                        return null
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error in overrideMediaItemBuilder: ${e.message}")
            null
        }
    }

    override fun shouldDisableCache(source: Source): Boolean {
        return false
    }

    override fun overrideMediaSourceFactory(
        source: Source,
        mediaSourceFactory: MediaSource.Factory,
        mediaDataSourceFactory: DataSource.Factory
    ): MediaSource.Factory? {
        return null
    }

    // ✅ FIXED: Use SAME logic as OfflineDataSourceProvider
    fun isContentDownloaded(uri: String): Boolean {
        return try {
            // Periodic cache cleanup
            val currentTime = System.currentTimeMillis()
            if (currentTime - lastCacheCleanup > cacheCleanupInterval) {
                contentDownloadCache.clear()
                lastCacheCleanup = currentTime
                Log.d(TAG, "🧹 Cleaned up content cache")
            }

            // Check cache first
            contentDownloadCache[uri]?.let {
                Log.d(TAG, "📋 Cache hit for $uri: $it")
                return it
            }

            // ✅ CRITICAL FIX: Force cache refresh by calling data source provider directly
            val isDownloaded = OfflineVideoRegistry.getAppContext()?.let { context ->
                Log.d(TAG, "🔄 Force checking cache via DataSourceProvider for: $uri")
                OfflineDataSourceProvider.getInstance(context).isContentCached(uri)
            } ?: false

            // ✅ Update cache with correct result
            contentDownloadCache[uri] = isDownloaded
            Log.d(TAG, "💾 Content $uri is downloaded: $isDownloaded (updated cache)")

            isDownloaded
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error checking if content is downloaded: ${e.message}")
            false
        }
    }

    private fun extractHeadersFromSource(source: Source): Map<String, String>? {
        return try {
            val headersMap = mutableMapOf<String, String>()

            try {
                // ✅ More defensive reflection
                val headersField = source.javaClass.getDeclaredField("headers")
                if (headersField != null) {
                    headersField.isAccessible = true
                    @Suppress("UNCHECKED_CAST")
                    val headers = headersField.get(source) as? Map<String, String>
                    headers?.let {
                        if (it.isNotEmpty()) {
                            headersMap.putAll(it)
                            Log.d(TAG, "📋 Extracted ${it.size} headers from source")
                        }
                    }
                }
            } catch (e: NoSuchFieldException) {
                Log.d(TAG, "Headers field not found in source (normal for some sources)")
            } catch (e: SecurityException) {
                Log.d(TAG, "Security exception accessing headers (normal)")
            } catch (e: IllegalAccessException) {
                Log.d(TAG, "Cannot access headers field (normal)")
            } catch (e: ClassCastException) {
                Log.d(TAG, "Headers field is not Map<String, String> type")
            }

            headersMap.ifEmpty { null }
        } catch (e: Exception) {
            Log.w(TAG, "Unexpected error extracting headers: ${e.message}")
            null
        }
    }

    private fun getStreamKeys(uri: String): List<StreamKey> {
        return try {
            val context = OfflineVideoRegistry.getAppContext() ?: return emptyList()
            val downloadManager = OfflineVideoDownloaderModule.getDownloadManager() ?: return emptyList()

            // ✅ Use timeout to prevent ANR
            val future = cacheCheckExecutor.submit<List<StreamKey>> {
                val downloadsCursor = downloadManager.downloadIndex.getDownloads()
                var streamKeys = emptyList<StreamKey>()

                try {
                    while (downloadsCursor.moveToNext()) {
                        val download = downloadsCursor.download
                        if (download.request.uri.toString() == uri &&
                            download.state == Download.STATE_COMPLETED) {
                            streamKeys = download.request.streamKeys
                            break
                        }
                    }
                } finally {
                    downloadsCursor.close()
                }

                streamKeys
            }

            // ✅ Wait max 500ms for stream keys
            future.get(500L, TimeUnit.MILLISECONDS)

        } catch (e: TimeoutException) {
            Log.w(TAG, "⏰ Stream keys retrieval timed out for: $uri")
            emptyList()
        } catch (e: Exception) {
            Log.w(TAG, "❌ Stream keys retrieval failed: ${e.message}")
            emptyList()
        }
    }
}
